storage/file_persister.go

```diff
@@ -43,3 +49,144 @@ func (l *LocalFilePersister) Persist(path string, data io.Reader) (err error) {
 
 	return nil
 }
+
+// RemoteFilePersister is to be used when files created by the browser module need
+// to be uploaded to a remote location. This uses a preSignedURLGetterURL to
+// retrieve one pre-signed URL. The pre-signed url is used to upload the file
+// to the remote location.
+type RemoteFilePersister struct {
+	preSignedURLGetterURL string
+	headers               map[string]string
+	basePath              string
+
+	httpClient *http.Client
+}
+
+// NewRemoteFilePersister creates a new instance of RemoteFilePersister.
+func NewRemoteFilePersister(pURL string, h map[string]string, bp string) *RemoteFilePersister {
+	return &RemoteFilePersister{
+		preSignedURLGetterURL: pURL,
+		headers:               h,
+		basePath:              bp,
+		httpClient: &http.Client{
+			Timeout: time.Second * 10,
+		},
+	}
+}
+
+// Persist will upload the contents of data to a remote location.
+func (r *RemoteFilePersister) Persist(ctx context.Context, path string, data io.Reader) (err error) {
+	pURL, err := r.getPreSignedURL(ctx, path)
+	if err != nil {
+		return fmt.Errorf("getting presigned url: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, pURL, data)
+	if err != nil {
+		return fmt.Errorf("creating upload request: %w", err)
+	}
+
+	resp, err := r.httpClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("performing upload request: %w", err)
+	}
+	defer resp.Body.Close() //nolint:errcheck
+
+	if _, err := io.Copy(io.Discard, resp.Body); err != nil {
+		return fmt.Errorf("draining upload response body: %w", err)
+	}
+
+	if err := checkStatusCode(resp); err != nil {
+		return fmt.Errorf("uploading status code out of range: %w", err)
+	}
+
+	return nil
+}
+
+func checkStatusCode(resp *http.Response) error {
+	if resp.StatusCode == http.StatusTooManyRequests {
+		return fmt.Errorf("request is being rate limited")
+	}
+
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
+		return fmt.Errorf("status code is %d", resp.StatusCode)
+	}
+
+	return nil
+}
+
+// getPreSignedURL will retrieve the presigned url for the current file.
+func (r *RemoteFilePersister) getPreSignedURL(ctx context.Context, path string) (string, error) {
+	b, err := buildPresignedRequestBody(r.basePath, path)
+	if err != nil {
+		return "", fmt.Errorf("building request body: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, r.preSignedURLGetterURL, bytes.NewReader(b))
+	if err != nil {
+		return "", fmt.Errorf("creating request: %w", err)
+	}
+
+	for k, v := range r.headers {
+		req.Header.Add(k, v)
+	}
+
+	resp, err := r.httpClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("performing request: %w", err)
+	}
+	defer resp.Body.Close() //nolint:errcheck
+
+	if err := checkStatusCode(resp); err != nil {
+		return "", fmt.Errorf("status code out of range: %w", err)
+	}
```

Comment:
I believe we can remove this error wrapping to reduce noise. In that case, it makes sense to return the second error in `checkStatusCode` like this:

```go
return fmt.Errorf("server returned %d (%s)", resp.StatusCode, http.StatusText(resp.StatusCode))
```

The error will be clearer and less noisy:
- getting presigned url: https://..: server returned 503 (Service Unavailable)
- persisting: server returned 429 (too many requests)

Instead of:
- getting presigned url: https://..: status code out of range: status code is 503
- uploading status code out of range: status code is 429

WDYT?

Reply:
Yep, i think this makes sense. I'm not so sure about adding the text versions of the http status codes, and instead will just work with the http status codes.

Reply:
Resolved in ac20d5b3d532c72443b9d5648c925b77b56f3a4d

Reply:
Nice. The text was about seeing what the status code means for users and us, so nobody needs to look it up.

Reply:
Maybe it wasn't clear in my previous comment, but, I also suggested combining these status code into one. Instead of saying, "request is being rate limited", "too many requests" is already clear. That's also why I suggest using status texts.

Reply:
Sure resolved in b9b25537c7f2cddb5c375c5dfa0301113ead217a